File: python_src/19.删除链表的倒数第-n-个结点.py
#
# @lc app=leetcode.cn id=19 lang=python3
#
# [19] 删除链表的倒数第 N 个结点
#
from typing import *

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next: ListNode = next

    def __str__(self) -> str:
        return f"{self.val}->{self.next}"

    @staticmethod
    def create(l: List[int]) -> "ListNode":
        head = ListNode(None)
        cur = head
        for i in l:
            cur.next = ListNode(i)
            cur = cur.next
        return head.next

class Solution0:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if head.next is None: return None

        head = ListNode(None, head)
        lst: List[ListNode] = []
        cur = head
        while cur:
            lst.append(cur)
            cur = cur.next
        lst[-n - 1].next = lst[-n].next

        return head.next

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if head.next is None: return None

        head = ListNode(None, head)
        l, r = head, head
        gap = 0
        while gap <= n:
            r = r.next
            gap += 1

        while r:
            l = l.next
            r = r.next

        l.next = l.next.next
        return head.next
# @lc code=end

s = Solution()
print(s.removeNthFromEnd(ListNode.create([1,2,3,4,5]), 2))
print(s.removeNthFromEnd(ListNode.create([1]), 1))
print(s.removeNthFromEnd(ListNode.create([1,2]), 1))
print(s.removeNthFromEnd(ListNode.create([1,2]), 2))

File: rust_src/19.删除链表的倒数第-n-个结点.rs
/*
 * @lc app=leetcode.cn id=19 lang=rust
 *
 * [19] 删除链表的倒数第 N 个结点
 */

// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn remove_nth_from_end(head: Option<Box<ListNode>>, n: i32) -> Option<Box<ListNode>> {
		let mut root = Some(Box::new(ListNode::new(0)));

		root.as_mut().unwrap().next = head;
		let mut end = root.as_ref().unwrap();

		for _ in 0..n
		{
			end = end.next.as_ref().unwrap();
		}
		
		let mut tmp = root.clone();
		let mut cur = tmp.as_mut().unwrap();
		
		while end.next.is_some()
		{
			end = end.next.as_ref().unwrap();
			cur = cur.next.as_mut().unwrap();
		}

		let node = cur.next.take().unwrap();
		cur.next = node.next;

		return tmp.unwrap().next;
    }
}
// @lc code=end

